### PR TITLE
[FEAT] Perf Tuning + Functional API

### PR DESCRIPTION
**Pull request by pzuraq**
This adds the perf tuning and functional API changes made in the 0.38
release to master, from #963

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -87,6 +87,10 @@
     "tslint-config-prettier": "^1.15.0",
     "tslint-plugin-prettier": "^2.0.1"
   },
+  "resolutions": {
```
Comment (rwjblue) [inline package.json]: Why do we need these resolutions? When can we remove them (maybe make an issue to track that)?

Comment (pzuraq) [reply package.json]: TypeScript doesn’t allow us to cast `any` to type `unique symbol` until more recently, and its the only way we can polyfill Symbol and type it correctly. We should likely update to the latest TS when possible (brocolli-typescript-transpiler needs to update). There are some other breaking changes on latest though

Comment (pzuraq) [reply package.json]: I think https://github.com/glimmerjs/glimmer-vm/issues/946 kind of tracks this

Comment (pzuraq) [reply package.json]: Opened #965

**Pull request merged**